### PR TITLE
Add experiment id helpers and generation speed logging

### PR DESCRIPTION
**Pull request by mmcdermott**
## Summary
- create `GenerationSpeedLogger` callback to log predict speed
- allow re-use of MLFlow/WandB run ids via helper functions
- integrate helpers into training and generation scripts
- register new callback only for generation runs
- fix Hydra defaults order for generation config
- add dedicated `generate` trainers and document logger helpers

## Testing
- `ruff check .`
- `pytest -q` *(failed: KeyboardInterrupt)*

------
https://chatgpt.com/codex/tasks/task_e_684196455518832c8f64c35153699a39

### CODE REVIEW

**Comment by mmcdermott** [comment]
## Rebase + review

Rebased this branch onto the latest `dev` (equivalent to current `main` after #78) and fixed three blockers so the PR is now in a mergeable state locally. Changed the base from `main` → `dev` to match the repo's feature-branch flow.

### Blockers found and fixed

1. **`README.md` doctest (top-level)** — the `print_directory("./src/MEDS_EIC_AR/configs", ...)` doctest asserts an exact directory listing and was missing the four new yaml files introduced by this PR (`generation.yaml`, `generation_speed_logger.yaml`, `demo_generate.yaml`, `generate.yaml`). Updated the expected tree.
2. **`src/MEDS_EIC_AR/configs/README.md` doctest** — identical issue, same four files missing from the expected listing. Updated.
3. **`save_logger_run_ids` docstring example** — used an unsupported combination of parenthesized-`with` + backslash-continuation inside a doctest, which raised `SyntaxError` under pytest doctest collection. It also passed an empty string `""` to `yaml_disk`, which parses as `None` and raises `TypeError`. Rewrote the example to name the two `patch(...)` context managers as variables and use `tempfile.TemporaryDirectory` directly.

With these fixes `uv run pytest -q` is green locally (47 passed) on the rebased branch.

### Assessment — is this PR helpful / ready?

**Helpful, and close to ready.** The three pieces land cleanly:

- `is_wandb_logger` / `is_mlflow_logger` helpers — small and self-contained, with `ImportError`-safe probes so they don't force a hard dep on wandb/mlflow.
- `apply_saved_logger_run_ids` / `save_logger_run_ids` — gives runs a way to resume a wandb/mlflow experiment id across `pretrain` → `generate_trajectories`, so both phases log into the same run. Writes to `<output_dir>/loggers/{mlflow_run_id,wandb_run_id}.txt`, reads them back when a downstream Hydra trainer config is being instantiated.
- `GenerationSpeedLogger` Lightning callback — logs `predict/avg_epoch_time_sec` at the end of prediction by diffing `time.perf_counter()` over `on_predict_epoch_{start,end}`. Addresses part of #53.

### Non-blocking observations (worth a follow-up but not required for merge)

- **`configs/trainer/callbacks/generation.yaml` inherits from `default`**, which pulls in `model_checkpoint`, `early_stopping`, and `learning_rate_monitor`. Those callbacks are no-ops during `trainer.predict()` (they only hook training events) but it would be cleaner to have generation start from a minimal base rather than inheriting training-only callbacks. Minor.
- **`GenerationSpeedLogger` granularity is coarse.** It only logs `avg_epoch_time_sec` on `on_predict_end`. Issue #53 asks for generation-speed telemetry that can be correlated against the pre-training run; `tokens/sec` and `trajectories/sec` would be more informative than epoch wall-clock, and ideally logged per-batch so a partial run still yields useful data. This is worth an expansion but doesn't block the basic wiring landing.
- **`save_logger_run_ids` writes `logger.experiment.id` for wandb** — depending on when this is called, `logger.experiment` may lazily create the underlying wandb run. Worth confirming that calling it in `generate_trajectories` after `trainer.predict()` doesn't spawn an extra run for an already-finished wandb logger. Integration test would help.

### Recommendation

Ready to merge into `dev` once CI confirms. The three fixes I pushed are trivial and don't touch the PR's actual logic — the original author's code is intact. Follow-ups for #53 (per-batch speed metrics, tokens/trajectories-per-second) can go in a separate PR.

File: src/MEDS_EIC_AR/utils.py [hunk 1]
```diff
@@ -355,3 +388,99 @@ def save_resolved_config(cfg: DictConfig, fp: Path) -> bool:
...
+        target = str(logger_cfg.get("_target_", "")).lower()
+        if "wandb" in target:
+            fp = log_dir / "wandb_run_id.txt"
+            if fp.is_file() and not logger_cfg.get("id"):
+                logger_cfg["id"] = fp.read_text().strip()
+                logger_cfg.setdefault("resume", "allow")
+        elif "mlflow" in target:
+            fp = log_dir / "mlflow_run_id.txt"
+            if fp.is_file() and not logger_cfg.get("run_id"):
+                logger_cfg["run_id"] = fp.read_text().strip()
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/utils.py]: Good catch — addressed in 1b9cdfd. `save_logger_run_ids` now also persists `tracking_uri` from the MLFlowLogger (via the private `_tracking_uri` attribute with a fallback to a public `tracking_uri` if a future Lightning version exposes one), and `apply_saved_logger_run_ids` restores it when the saved file is present and the current config hasn't explicitly set one. Explicit caller override wins — this preserves the escape hatch for `yes, I really do want to log into a new store`. Doctest expanded to cover both the default-restore path and the caller-override path.

File: src/MEDS_EIC_AR/training/callbacks.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        self._epoch_times.append(time.perf_counter() - self._epoch_start)
+
+    def on_predict_end(self, trainer, pl_module) -> None:
+        if not self._epoch_times:
+            return
+        avg_epoch_time = sum(self._epoch_times) / len(self._epoch_times)
+        metrics = {"predict/avg_epoch_time_sec": avg_epoch_time}
+        for logger in _ensure_logger_sequence(trainer.loggers):
+            logger.log_metrics(metrics, step=trainer.global_step)
+
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/training/callbacks.py:76]: Fixed in 1b9cdfd. `on_predict_end` now returns early if `trainer.is_global_zero` is false. `_epoch_times` is still measured per-rank (fine — each rank's wall-clock is independent), but only rank-zero writes the aggregate metric. Added a paired unit test (`test_generation_speed_logger_logs_once_on_rank_zero`) that exercises both the happy path and the non-global-zero path using a `FakeTrainer` + `RecordingLogger`, so any future regression here fails loudly in CI.

File: src/MEDS_EIC_AR/training/callbacks.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            return
+        self._epoch_times.append(time.perf_counter() - self._epoch_start)
+
+    def on_predict_end(self, trainer, pl_module) -> None:
+        if not self._epoch_times:
+            return
+        avg_epoch_time = sum(self._epoch_times) / len(self._epoch_times)
+        metrics = {"predict/avg_epoch_time_sec": avg_epoch_time}
+        for logger in _ensure_logger_sequence(trainer.loggers):
+            logger.log_metrics(metrics, step=trainer.global_step)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/training/callbacks.py:75]: Added in 1b9cdfd: `tests/test_logger_helpers.py::test_generation_speed_logger_logs_once_on_rank_zero`. Drives the full callback lifecycle (`on_predict_start` → `on_predict_epoch_start` → `on_predict_epoch_end` → `on_predict_end`) on a fake trainer against a recording logger, and asserts: (1) exactly one `log_metrics` call after a single predict run, (2) metric name is `predict/avg_epoch_time_sec` and value ≥ 0, (3) step argument matches `trainer.global_step`, (4) non-global-zero trainers produce zero calls. Covers the rank-zero gate from comment ② as a side effect.

File: tests/test_logger_helpers.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    def __init__(self, id="wandb_run"):
+        self.id = id
+
+
+class DummyWandbLogger:
+    def __init__(self, exp_id="wandb_run"):
+        self.experiment = DummyWandBExp(exp_id)
+
+
+def test_apply_saved_logger_run_ids(tmp_path, monkeypatch):
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_logger_helpers.py]: Dropped in 1b9cdfd. `test_apply_saved_logger_run_ids` is now `def test_apply_saved_logger_run_ids(tmp_path):` — no `monkeypatch` argument.

File: src/MEDS_EIC_AR/utils.py [hunk 5]
```diff
@@ -355,3 +388,134 @@ def save_resolved_config(cfg: DictConfig, fp: Path) -> bool:
...
+            # Restore ``tracking_uri`` too. An MLflow ``run_id`` is only resolvable against the
+            # tracking store it was created in — in this repo the default tracking_uri is derived
+            # from ``${log_dir}`` (i.e. the current run's ``output_dir``), so reusing a saved
+            # run_id in a fresh output dir without also restoring the original tracking_uri sends
+            # the resumed run to the wrong store (or 404s). If the caller explicitly set
+            # ``tracking_uri`` in the current config, we don't override it — that's the escape
+            # hatch for "yes, I really do want to log into a new store." See PR #73 review.
+            tracking_fp = log_dir / "mlflow_tracking_uri.txt"
+            if tracking_fp.is_file() and not logger_cfg.get("tracking_uri"):
+                logger_cfg["tracking_uri"] = tracking_fp.read_text().strip()
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/utils.py]: Good catch — fixed in 96caeaf by changing the rule. The restore is now gated on whether we just applied a saved `run_id`, not on whether `tracking_uri` happens to be empty: when the caller didn't explicitly set `run_id` (so we take the saved one), we also override `tracking_uri` with the saved value, regardless of what the current config had (including a default-interpolated one). The new rule means resumed runs always attach to the original store, and the "incoherent" case you flagged (resume run_id against a different store) is prevented by construction.

Escape hatch for "new run in a different store": set `run_id` explicitly in the config — that suppresses both the run_id restore and, by the same flag, the tracking_uri restore.

This changes the semantics from "caller-override of tracking_uri always wins" to "saved tracking_uri wins when we're resuming". That's a stronger contract but the right one — a user who explicitly overrides tracking_uri while leaving run_id unset was mixing two resume modes that can't both be true at once.

Doctest expanded to cover both the default-override and the explicit-run_id escape hatch.

File: tests/test_logger_helpers.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    (log_dir / "mlflow_run_id.txt").write_text("abc")
+    (log_dir / "wandb_run_id.txt").write_text("xyz")
+
+    utils.apply_saved_logger_run_ids(cfg, tmp_path)
+
+    assert cfg.loggers[0]["run_id"] == "abc"
+    assert cfg.loggers[1]["id"] == "xyz"
+    assert cfg.loggers[1]["resume"] == "allow"
+
+
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_logger_helpers.py:44]: Added in 96caeaf: `test_apply_saved_logger_run_ids_overrides_default_mlflow_tracking_uri` — locks in that a populated current `tracking_uri` is overridden by the saved one when we apply a saved `run_id`. Also added a paired `test_apply_saved_logger_run_ids_preserves_explicit_run_id` that covers the escape hatch: an explicit `run_id` in the current config disables both the run_id and tracking_uri restores, so a caller who wants a fresh run in a new store gets exactly that.

Used a cleaner tmp-based path (`str(tmp_path / ...)`) rather than the `${log_dir}/...` interpolation in your suggested snippet so the test is self-contained and doesn't depend on OmegaConf resolver wiring — the helper operates on post-interpolation config values in production anyway, so the test setup mirrors what the helper actually sees.

File: src/MEDS_EIC_AR/training/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        if not self._epoch_times:
+            return
+        # Rank-gate: in distributed predict, every rank hits ``on_predict_end``. Writing
+        # ``log_metrics`` from every rank produces duplicated metric writes and races some
+        # backends (e.g., MLflow), so only rank-zero logs. ``_epoch_times`` measurement itself
+        # is per-rank and independent, which is fine — we just restrict the *write*.
+        if not trainer.is_global_zero:
+            return
+        avg_epoch_time = sum(self._epoch_times) / len(self._epoch_times)
+        metrics = {"predict/avg_epoch_time_sec": avg_epoch_time}
```
Comment (mmcdermott) [inline src/MEDS_EIC_AR/training/callbacks.py]: what about standard deviation or other measures of variance? Or max or min?

**Review by mmcdermott — COMMENTED** [review]
Is there a way to do an integration test with real wandb and MLFlow loggers, in files gated by installation of those packages, and then tested on CI in distinct stages that additionally install those dependencies -- and ideally without actually pinging remote servers for either? That would be ideal, if possible.

Also, will either of these changes have any performance implications for the model (in runtime, e.g.)? Or reliability (e.g., it will crash on re-run if we have something wrong in the run id application logic)?

**Comment by mmcdermott** [comment]
additional comments from chatGPT (from one commit prior): 
Yes. My take is: **this looks good now, and I’d lean merge/approve**, with a couple of non-blocking follow-ups.

What changed my view is that the branch now handles the originally scary MLflow case correctly. The repo’s default MLflow logger config derives `tracking_uri` from the current run’s `log_dir`, so merely restoring a saved `run_id` would be wrong if generation runs in a different output directory. Lightning’s MLflow logger treats `run_id` and `tracking_uri` as separate configuration fields, and the current PR now restores the saved `tracking_uri` whenever it restores a saved `run_id`, rather than only when `tracking_uri` is absent. It also has regression tests for both “override the default tracking URI on resume” and “preserve an explicitly set fresh run_id/tracking_uri.” That removes what would otherwise have been the main correctness blocker. ([GitHub][1])  

The overall wiring is also sensible: it applies saved logger IDs before trainer instantiation in both `pretrain` and `generate_trajectories`, then persists them again after the run. That is the right place to do it if the goal is “generation logs back into the original experiment record.” 

The `GenerationSpeedLogger` part is also in decent shape now. The callback records predict-time wall clock, logs only on rank zero, and there is a unit test covering both the normal path and the non-global-zero path, which addresses the earlier distributed logging concern.  

The two things I’d still call out are both follow-up material, not merge blockers.

First, the metric is **pretty coarse**. Right now it only emits `predict/avg_epoch_time_sec`. That is better than nothing, but for actual generation-performance analysis it is a weak observable; `tokens/sec`, `rows/sec`, or `trajectories/sec`, ideally per batch or at least per predict pass, would be much more actionable. The PR discussion already points in that direction. ([GitHub][2])

Second, the generation callback bundle still appears to inherit from the default callback set, which means generation drags along training-only callbacks like checkpointing, early stopping, and LR monitoring. Those should be inert under `trainer.predict()`, so I do not think it is harmful, but it is a slightly muddy abstraction boundary. A minimal generation-only callback set would be cleaner. ([GitHub][2])

So the comment I’d leave is roughly:

> I think this is in good shape now. The important MLflow resume issue is handled correctly: restoring a saved `run_id` now also restores the saved `tracking_uri`, which is necessary because the repo default derives `tracking_uri` from the current run’s `log_dir`. The added tests for default-tracking-URI override and explicit fresh-run override are especially helpful.
>
> The rank-zero fix and unit test for `GenerationSpeedLogger` also make that part much safer.
>
> My remaining feedback is non-blocking: (1) `predict/avg_epoch_time_sec` is useful but still a coarse metric relative to the real throughput quantities we probably care about, and (2) generation still seems to inherit training-only callbacks from the default callback set, which is harmless but a bit untidy.
>
> Net: I’d be comfortable merging this and treating finer-grained generation telemetry / callback cleanup as follow-up work.

If you want, I can turn that into a tighter GitHub-ready review comment.

[1]: https://raw.githubusercontent.com/mmcdermott/MEDS_EIC_AR/dev/src/MEDS_EIC_AR/configs/trainer/logger/mlflow.yaml "raw.githubusercontent.com"
[2]: https://github.com/mmcdermott/MEDS_EIC_AR/pull/73 "Add experiment id helpers and generation speed logging by mmcdermott · Pull Request #73 · mmcdermott/MEDS_EIC_AR · GitHub"

File: src/MEDS_EIC_AR/training/callbacks.py (hunk 7 above)
Comment (mmcdermott) [reply src/MEDS_EIC_AR/training/callbacks.py]: Good point — added in 56ba646. `GenerationSpeedLogger` now also collects per-batch times (via `on_predict_batch_start/end`) and emits mean/min/max/std for both epoch- and batch-grained distributions, plus `total_time_sec` and `num_batches`.

I leaned into **per-batch** as the primary variance surface since batch timings are where the distribution actually shows up under rolling generation (sequence-length sensitivity + GPU warm-up + device contention). The per-epoch stats are cheap to compute from the same timer scaffolding, so both are reported; for a single-epoch predict call the epoch std collapses to 0 and the signal is in the batch-level numbers.

Used `statistics.pstdev` instead of `stdev` so the single-sample edge case yields a defined `0.0` rather than raising. The test now drives two batches and locks in the 10-key metrics set, monotonic min≤mean≤max, and the single-epoch std=0 collapse.

File: src/MEDS_EIC_AR/utils.py [hunk 8]
```diff
@@ -355,3 +388,163 @@ def save_resolved_config(cfg: DictConfig, fp: Path) -> bool:
...
+    log_dir = Path(run_dir) / "loggers"
+
+    for logger_cfg in loggers:
+        target = str(logger_cfg.get("_target_", "")).lower()
+        if "wandb" in target:
+            fp = log_dir / "wandb_run_id.txt"
+            if fp.is_file() and not logger_cfg.get("id"):
+                logger_cfg["id"] = fp.read_text().strip()
+                logger_cfg.setdefault("resume", "allow")
+        elif "mlflow" in target:
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/utils.py]: Good catch — fixed in 56ba646. Added a normalization pass at the top of `apply_saved_logger_run_ids` that:

1. Treats `trainer.logger` as a single entry (which may be a mapping, bool, None, or anything else).
2. Treats `trainer.loggers` as either a single mapping OR a sequence (list/ListConfig) — extending when it's a sequence, appending as a single entry otherwise.
3. Filters out anything without a `.get` method, which catches bools, None, and concrete `Logger` instances (when the helper is called on a post-instantiation config).

This matches Lightning's `logger: bool | Logger | Iterable[Logger]` contract. Added `test_apply_saved_logger_run_ids_handles_disabled_logger` covering four variants: `logger=false`, `logger=null`, `loggers=false`, and `loggers=[null, {real mlflow config}]` — the last one verifies the mixed case still patches the real entry while ignoring the null.

File: src/MEDS_EIC_AR/__main__.py [hunk 9]
```diff
@@ -290,4 +295,5 @@ def generate_trajectories(cfg: DictConfig):
             pa_table = GeneratedTrajectorySchema.align(predictions_df.to_arrow())
             pq.write_table(pa_table, out_fp)
 
+    save_logger_run_ids(trainer.loggers, Path(cfg.model_initialization_dir))
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/__main__.py]: Good catch — fixed in 56ba646 using your suggested `Path(cfg.output_dir)`. This mirrors what pretrain already does (saves into its own `output_dir`) and keeps the training run's save-point frozen.

The behavior I now have:
- pretrain saves into its `output_dir` (training run's save-point)
- generate_trajectories loads saved ids via `apply_saved_logger_run_ids(cfg.trainer, cfg.model_initialization_dir)` — still reads from the training dir
- generate_trajectories saves new ids via `save_logger_run_ids(..., cfg.output_dir)` — writes to its own output dir

Result: a user running the escape hatch (explicit `run_id=fresh` for a new-store generation) lands the fresh run's ids in their `output_dir`, and the training save-point at `model_initialization_dir/loggers/` stays untouched for future resumes.

**Review by mmcdermott — COMMENTED** [review]
Both asks addressed in e54a1a9.

## Integration tests

Added `tests/test_logger_helpers_integration.py` with four tests driving **real** `MLFlowLogger` and `WandbLogger` instances end-to-end:

1. `test_mlflow_save_and_apply_round_trip` — creates a real MLflow run against a local `file://` tracking URI, saves its id+uri via `save_logger_run_ids`, then `apply_saved_logger_run_ids` against a fresh config pointing at a different `file://` URI; asserts the saved uri wins.
2. `test_wandb_save_and_apply_round_trip` — creates a real offline WandbLogger, saves its `experiment.id`, applies to a fresh config; asserts id + `resume=allow`.
3. `test_is_logger_detectors_match_real_classes` — exercises `is_mlflow_logger` / `is_wandb_logger` against the real classes (not just the dummy classes from the unit tests). This catches upstream Lightning class-rename regressions.
4. `test_corrupted_saved_id_files_do_not_crash` — zero-byte / whitespace-only saved-id files must be treated as "no saved id", not restored as `run_id=""`.

**Offline by design**: MLflow uses a `file://` tracking URI, WandB uses `WandbLogger(offline=True)` + `WANDB_MODE=offline`. No network contact, no credentials needed.

**Gated by install**: the whole module is guarded by `pytest.importorskip("mlflow")` / `("wandb")`, so on a default install these collect-and-skip cleanly.

**Distinct CI stage**: added a new `run_integration_tests_with_optional_extras` job in `.github/workflows/tests.yaml` that installs the extras via the setup action's new `extras` input and runs just the integration tests. The main `run_tests_ubuntu` job stays fast — integration tests skip there (importorskip) rather than causing a dependency-install time penalty on every PR.

## Performance

Negligible. Concretely:

- `save_logger_run_ids`: one `logger.run_id` / `logger.experiment.id` attribute read per logger + one `write_text` per logger. Runs once at end-of-training and end-of-generation. Well under a millisecond total.
- `apply_saved_logger_run_ids`: one `is_file()` check + one `read_text()` per logger. Runs once before trainer instantiation. Also well under a millisecond.
- `GenerationSpeedLogger`: one `time.perf_counter()` pair per batch (and per epoch). That's ~100ns per call on x86; over a full generation run (say 10k batches) that's ~2ms total. Not observable next to model forward passes.

None of this is on any hot path, and none of it scales with model size or batch size.

## Reliability

Two questions here: "does our own logic crash on re-run", and "does a corrupted save-point brick future runs".

**Our own logic**: exercised by the unit tests (`test_apply_saved_logger_run_ids*`, 4 variants including bool/null/mixed-list/explicit-run_id-escape-hatch) plus the 4 new integration tests. Every code path through `apply_saved_logger_run_ids` has coverage, and the normalization pass (`hasattr(..., "get")` guard) catches any non-mapping config entry before it reaches `.get(...)`.

**Corrupted save-point**: hardened in e54a1a9 via a new `_read_saved_id(fp)` helper that is the single entry point for every `run_id.txt` / `tracking_uri.txt` read. It returns `None` for:
- missing files,
- files that raise `OSError` on read (permission-denied, concurrent-delete, disk failures — these now log a warning and fall through),
- files that strip to the empty string.

The upstream effect: a zero-byte `mlflow_run_id.txt` left over from a previously-failed pretrain no longer poisons the next generation run by restoring `run_id=""`. The restore silently no-ops, and the run proceeds with the pre-restore config. The new `test_corrupted_saved_id_files_do_not_crash` integration test locks this in for both MLflow and WandB.

**Comment by mmcdermott** [comment]
Addressed in e54a1a9 (integration tests + defensive saved-id reads) and 56ba646 (variance/std/min/max metrics). The two non-blocking follow-ups ChatGPT flagged — richer throughput metrics (tokens/sec / rows/sec / trajectories/sec) and trimming training-only callbacks from the generation bundle — are filed as issues #119 and #120 respectively.

**Comment by mmcdermott** [comment]
additional chatgpt comments:
Yes — **this now looks mergeable to me**, with only minor follow-up items.

What materially improved since the last pass is:

* The PR no longer writes generation logger IDs back into `model_initialization_dir`; it now saves them into the generation `output_dir`, which avoids clobbering the training run’s saved IDs when generation intentionally uses a fresh run. That was the last real correctness trap I was worried about. 
* `apply_saved_logger_run_ids` is now more robust about Hydra/Lightning logger shapes: it normalizes `trainer.logger` / `trainer.loggers`, skips non-mapping entries like `false`/`null`, and won’t blow up on disabled logging configs. 
* Saved-ID reads are now defensive: missing, unreadable, empty, or whitespace-only files are treated as “no saved ID” rather than poisoning the next run with something like `run_id=""`. That substantially improves re-run reliability. 
* The PR now has real offline integration tests for both MLflow and WandB, gated behind optional extras and run in a separate CI job. That is exactly the right way to test this surface. 
* `GenerationSpeedLogger` now reports richer timing summaries: total time, batch count, and mean/min/max/std at both epoch and batch granularity, instead of only a single average epoch time. 

So my current recommendation is:

> I think this is in good shape now and I’d be comfortable approving it.
>
> The remaining correctness/reliability concerns have been addressed: generation no longer overwrites the training run’s saved logger IDs, saved-ID restore handles disabled/non-mapping logger configs, and corrupted/blank saved-ID files now no-op instead of poisoning reruns. The addition of real offline MLflow/WandB integration tests plus a separate optional-extras CI lane is especially strong.
>
> I also like the improved timing telemetry: batch- and epoch-level mean/min/max/std plus total time and batch count is a meaningful step up from the earlier single average metric.
>
> My remaining comments are non-blocking only: there still aren’t direct throughput metrics like tokens/sec or trajectories/sec in this PR, and the generation callback bundle still appears to inherit training-oriented callbacks from the default set. Both feel fine as follow-up issues rather than reasons to hold this PR.

So: **approve / merge pending CI**, not “request changes.”

File: src/MEDS_EIC_AR/utils.py [hunk 10]
```diff
@@ -355,3 +388,201 @@ def save_resolved_config(cfg: DictConfig, fp: Path) -> bool:
...
+    # ``_target_`` resolution — though this helper usually runs *before* instantiation, it can
+    # also be called on already-resolved configs) is skipped silently.
+    raw_entries: list = []
+    if "logger" in trainer_cfg:
+        raw_entries.append(trainer_cfg.logger)
+    if "loggers" in trainer_cfg:
+        loggers_field = trainer_cfg.loggers
+        if isinstance(loggers_field, DictConfig | dict):
+            raw_entries.append(loggers_field)
+        elif isinstance(loggers_field, list | ListConfig):
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/utils.py:488]: Pushing back — this is a false positive. PEP 604's `X | Y` union syntax is supported by `isinstance` since Python 3.10 (``requires-python = ">=3.11"`` in this repo's `pyproject.toml`). Verified locally on Python 3.12.3:

```
>>> from omegaconf import DictConfig, ListConfig
>>> isinstance(DictConfig({'a': 1}), DictConfig | dict)
True
>>> isinstance([1, 2, 3], list | ListConfig)
True
```

CI also passes the tests that exercise this path (`test_apply_saved_logger_run_ids_handles_disabled_logger`, which hits both branches). The code is correct as-is; switching to the tuple form would just be a style regression back toward pre-3.10 idioms.

Reference: [CPython 3.10 What's New — PEP 604](https://docs.python.org/3/whatsnew/3.10.html#pep-604-new-type-union-operator) ("`isinstance` and `issubclass` now accept a union of types as parameter").

**Comment by mmcdermott** [comment]
Thanks — glad the current round landed well. Confirming the two remaining non-blocking items are captured:

- **Throughput metrics (tokens/sec, rows/sec, trajectories/sec)**: #119
- **Trim training-only callbacks from the generation bundle**: #120

Happy to land those as separate PRs after this one merges.
